C/eval.c

```diff
@@ -587,6 +597,35 @@ static bool runTCO(evalState state, call* stack, const dag_node* dag, type* type
     }
   }
   assert(pc == len);
+
+  return true;
+}
+
+/* Inspects the stack contents after a successful runTCO execution to verify anti-DOS properties:
+ * 1. All dag nodes were executed at least once.
+ * 2. Both branches of every CASE node were executed.
+ *
+ * If these are violoated it means that the dag should have had unused nodes prune or should have transformed
+ * CASE node with unused branches into either ASSERTL or ASSERTR nodes.
+ *
+ * Returns false if any of the anti-DOS checks fail.  Otherwise returns true.
+ *
+ * Precondition: call stack[len];
+ *               dag_node dag[len];
+ */
+static bool antiDos(const call* stack, const dag_node* dag, size_t len) {
+  for(size_t i = 0; i < len; ++i) {
+    /* All combinators must be executed at least once. */
+    flags_type test_flags = EXEC_FLAG;
+
+    /* Both branches of every case combinator must be executed at least once. */
+    if (CASE == dag[i].tag) test_flags |= CASE_LEFT_FLAG || CASE_RIGHT_FLAG;
```

Comment:
In 2d58ec81f4a848b7f7083e3e85ba85e78545b7b6:

Unless I'm crazy this code is just wrong -- it sets `test_flags` to the *logical or* rather than *bitwise or* of the two CASE_*_FLAGS. But then even if you use a bitwise or, the result is to set both flags on, and then the `if` below will check that *both* children were executed. But in the case that one of the children is an ASSERT, this check is incorrect.

Reply:
You are not crazy.

Reply:
Fixed.